src/sinks/aws_cloudwatch_logs/request.rs

```diff
@@ -0,0 +1,178 @@
+use super::CloudwatchError;
+use futures::{sync::mpsc, Future, Poll};
+use rusoto_core::RusotoFuture;
+use rusoto_logs::{
+    CloudWatchLogs, CloudWatchLogsClient, CreateLogStreamError, CreateLogStreamRequest,
+    DescribeLogStreamsError, DescribeLogStreamsRequest, DescribeLogStreamsResponse, InputLogEvent,
+    PutLogEventsError, PutLogEventsRequest, PutLogEventsResponse,
+};
+
+pub struct CloudwatchFuture {
+    client: CloudWatchLogsClient,
+    state: State,
+    stream_name: String,
+    group_name: String,
+    events: Option<Vec<InputLogEvent>>,
+    token: Option<String>,
+    token_tx: mpsc::Sender<Option<String>>,
+}
+
+enum State {
+    Idle,
+    Token(Option<String>),
+    CreateStream(RusotoFuture<(), CreateLogStreamError>),
+    DescribeStream(RusotoFuture<DescribeLogStreamsResponse, DescribeLogStreamsError>),
+    Put(RusotoFuture<PutLogEventsResponse, PutLogEventsError>),
+}
+
+impl CloudwatchFuture {
+    pub fn new(
+        client: CloudWatchLogsClient,
+        stream_name: String,
+        group_name: String,
+        events: Vec<InputLogEvent>,
+        token: Option<String>,
+        token_tx: mpsc::Sender<Option<String>>,
+    ) -> Self {
+        Self {
+            client,
+            events: Some(events),
+            state: State::Idle,
+            stream_name,
+            group_name,
+            token,
+            token_tx,
+        }
+    }
+
+    fn put_logs(
+        &mut self,
+        sequence_token: Option<String>,
+        log_events: Vec<InputLogEvent>,
+    ) -> RusotoFuture<PutLogEventsResponse, PutLogEventsError> {
+        let request = PutLogEventsRequest {
+            log_events,
+            sequence_token,
+            log_group_name: self.group_name.clone(),
+            log_stream_name: self.stream_name.clone(),
+        };
+
+        self.client.put_log_events(request)
+    }
+
+    fn describe_stream(
+        &mut self,
+    ) -> RusotoFuture<DescribeLogStreamsResponse, DescribeLogStreamsError> {
+        let request = DescribeLogStreamsRequest {
+            limit: Some(1),
+            log_group_name: self.group_name.clone(),
+            log_stream_name_prefix: Some(self.stream_name.clone()),
+            ..Default::default()
+        };
+
+        self.client.describe_log_streams(request)
+    }
+
+    fn create_log_stream(&mut self) -> RusotoFuture<(), CreateLogStreamError> {
+        let request = CreateLogStreamRequest {
+            log_group_name: self.group_name.clone(),
+            log_stream_name: self.stream_name.clone(),
+        };
+
+        self.client.create_log_stream(request)
+    }
+}
+
+impl Future for CloudwatchFuture {
+    type Item = ();
+    type Error = CloudwatchError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        macro_rules! try_ready {
```

Comment:
This might be a little confusing for readers already familiar with the normal `try_ready`